src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java

```diff
@@ -42,12 +42,16 @@ public PathFragment getEffectiveSourcePath(Artifact source) {
   @Override
   public void finalizeCompileActionBuilder(
       RuleContext ruleContext, CppCompileActionBuilder actionBuilder) {
-    actionBuilder.setCppConfiguration(ruleContext.getFragment(CppConfiguration.class));
+    CppConfiguration cppConfiguration = ruleContext.getFragment(CppConfiguration.class);
+    actionBuilder.setCppConfiguration(cppConfiguration);
     actionBuilder.setActionContext(CppCompileActionContext.class);
-    // Because Bazel does not support include scanning, we need the entire crosstool filegroup,
-    // including header files, as opposed to just the "compile" filegroup.
-    actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
-    actionBuilder.setShouldScanIncludes(false);
+    boolean shouldScanIncludes = cppConfiguration.shouldScanIncludes();
+    if (!shouldScanIncludes) {
+      // When not performing include scanning, we need the entire crosstool filegroup,
+      // including header files, as opposed to just the "compile" filegroup.
+      actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
+    }
+    actionBuilder.setShouldScanIncludes(shouldScanIncludes);
```

Comment:
@damienmg Out of curiousity, is scanning of includes using the same `genrule` infrastructure we have available in Skylark or is it cheating?